[CL-610] Implement reference distributions for representativeness

adessy opened the pull request:
(CL-610)

This PR introduces a new resource `RefDistribution` that will be used to store census/demographic data. A reference distribution is always tied to a user custom field, such as gender and neighbourhood, and describes how the total population of the city (or target population in general) is distributed among the different options of the custom field. In essence, something like: 
```json
{
  "male": 13500,
  "female": 14000
}
```

This new resource and endpoints is the first in allowing the front-end to build the representativeness dashboard, and more specifically this type of chart:
![image](https://user-images.githubusercontent.com/209268/167091270-3111eadf-24ae-4535-a48b-364fc3c6a206.png)

## A few comments...

- We are doing a bit of back and forth on the mathematical definition of the representativeness score. Some approaches just need the distribution (percentages or an unnormalised distribution), others require the actual population counts. That's the reason why `RefDistribution` might not be the most accurate name. That's what I started with and it's still kinda works even for population counts. So I'll stick with it for now.
- The `POST .../custom_fields/:id/reference_distribution` endpoint has a very non standard behaviour. When posting a new reference distribution to a custom field, it removes the existing one if any. In the next PRs, the plan is to keep reference distribution immutable (only way to change it is to create a new one), but to keep the older versions as well. Each representativeness score will hold a reference to the distribution that was used to compute it. 
- For the sake of simplicity and bc I'm not satisfied with the way we specify dependencies between engines, I decided to make 'representativeness' part of the `user_custom_fields` engine.
- I had a holiday break in the middle of this PR, so pretty sure I forgot some things 🙈

adessy committed:
CL-610 Create DB table for reference distributions

adessy committed:
CL-610 Add minimal model for reference distributions

It also defines a factory for ref distribution and add it to the factory path.

adessy committed:
CL-610 Add (draft) POST and GET endpoints for reference distributions

adessy committed:
CL-610 Add the DELETE endpoint for reference distributions

adessy committed:
CL-610 Remove endpoints not yet implemented for ref. distributions

adessy committed:
CL-610 Implement validations for reference distributions

adessy committed:
CL-610 Test "not found" case for GET .../reference_distribution

adessy committed:
CL-610 Rework endpoints for reference distribution

POST now replaces the existing reference distribution if any. In the future, POST will just create a new distribution without deleting the previous ones. (Reference distribution would be immutable resources. We'd keep the former distributions bc each representativeness score will keep a reference to the distribution that was used for its computations.)

This commit also:
- adds logic for side effects
- ensures that the endpoint returns 422 when the distribution is invalid

adessy committed:
CL-610 Add missing prefix for table names

adessy committed:
CL-610 Fix validation tests for reference distributions

adessy committed:
CL-610 Add the has-many-distributions association to custom fields

adessy committed:
CL-610 Add the feature flag for representativeness

adessy committed:
CL-610 Merge branch 'master' into CL-610/representativeness-reference-distribution

koendehondt commented on back/engines/commercial/user_custom_fields/app/controllers/user_custom_fields/web_api/v1/ref_distributions_controller.rb:
```diff
@@ -0,0 +1,55 @@
+          # Removes the existing reference distribution if any.
+          Representativeness::RefDistribution.where(
+            custom_field_id: params[:custom_field_id]
+          ).destroy_all
```
Instead of writing the comment `# Removes the existing reference distribution if any.`, consider extracting this code to a method called `remove_existing_reference_distribution_if_any(custom_field_id)`. Then the code speaks for itself and the comment is not required.

adessy replied:
I can agree, but I think it starts to be a bit on the edge in this case.

koendehondt replied:
> it starts to be a bit on the edge in this case.

What do you mean?

I use the following rule of thumb: if a comment is needed to explain a piece of code, abstraction is needed. Using expressive variable names and extracting code to new methods are the ways to achieve abstraction during day to day coding. By doing so, the code will be easier to read. The method names express what the methods do, methods will be shorter, and code comments are not necessary.

Note that I do not say here that all code comments are useless. I am just saying that by refactoring code into smaller methods with good names, most method level comments are not required anymore.

koendehondt commented on back/engines/commercial/user_custom_fields/app/controllers/user_custom_fields/web_api/v1/ref_distributions_controller.rb:
```diff
@@ -0,0 +1,55 @@
+      end
+
+      def reference_distribution
+        @reference_distribution ||= authorize(
```
Although the implementation is fine, I see that the pattern of using a `before_action` to set the model is not used, while elsewhere it is. For instance the `ProjectsController` has `before_action :set_project, only: %i[show update reorder destroy]`. So I wonder what the best practice is: using a `before_action`, or not.

adessy replied:
For this specific case, we are still very much relying on personal preference of the dev. Both tastes co-exist in the codebase and we usually don't frown too much about it. Personally, I find using a `before_action` more of an anti-pattern than anything else in this case. I agree there is always value in making the codebase more consistent, but you got to pick your battles, right? 🙂

koendehondt commented on back/engines/commercial/user_custom_fields/app/models/user_custom_fields/patches/custom_field.rb:
```diff
@@ -0,0 +1,14 @@
+module UserCustomFields::Patches::CustomField
+  def self.included(base)
+    base.class_eval do
+      # TODO: Refactor CustomFields to use STI and add the following only to user custom fields.
```
Is this to-do something that you like to do as part of this PR or as part of the Jira ticket, or is this a more general issue that requires a solution in the future? If the latter, please indicate that, otherwise it is not clear for the reader why this comment is here.

adessy replied:
Note very good practice, but it was more of a note to myself. (I will touch this file quite a few more times before the representativeness feature is done.)

I was quite not sure about doing it or not. So maybe you can help me making a decision 😛 I noticed while working on those files that:
- the OS refactoring of user custom fields (extraction from the core app to an engine) was not done properly,
- using STI would probably be a better / more elegant way to implement the different types of custom fields (user, form).

That's something I'd like to rework if time allows it, but I'm afraid of opening a pandora box. In any case, I didn't want to include this refactoring the current PR to keep it small and focused. What do you think?

koendehondt replied:
> the OS refactoring of user custom fields (extraction from the core app to an engine) was not done properly,

I like to know more about this, as idea custom fields (Sara's work) is something that the Methods Squad will dive into starting this week. I like to do a Tuple session with you to discuss your thoughts. Tomorrow at 9:00?

> That's something I'd like to rework if time allows it, but I'm afraid of opening a pandora box. In any case, I didn't want to include this refactoring the current PR to keep it small and focused. What do you think?

Keep this PR small and focused, I would say.

koendehondt commented on back/engines/commercial/user_custom_fields/app/models/user_custom_fields/representativeness/ref_distribution.rb:
```diff
@@ -0,0 +1,45 @@
+    distribution.values.sum
+  end
+
+  # TODO: better name (bc there are both counts and probabilities)
```
Please solve this to-do.

adessy replied:
I'm stumped on that one. I was thinking `description`, or maybe `specification` but not really satisfied by any of those. What do you think?

The method returns something like:
```ruby
{
  "7e49c030-14e0-4128-b2c5-22e50f0120ad"=>{:count=>42, :probability=>0.6774193548387096},
  "5d5f64c6-b973-4d2b-b9be-682bdfe2ab6f"=>{:count=>20, :probability=>0.3225806451612903}
}
```

koendehondt replied:
It is a bit straightforward, but `counts_and_probabilities` expresses exactly what it is.

adessy replied:
Sounds obvious now you say it. So, `counts_and_probabilities` it is.

koendehondt commented on back/engines/commercial/user_custom_fields/app/models/user_custom_fields/representativeness/ref_distribution.rb:
```diff
@@ -0,0 +1,45 @@
+
+    unless distribution.keys.to_set <= custom_field.custom_field_option_ids.to_set
+      errors.add(:distribution, 'options must be a subset of the options of the associated custom field.')
+    end
```
Consider refactoring this `unless` to a guard clause, just like the guard clauses on lines 38 and 39.

```suggestion
    return if distribution.keys.to_set >= custom_field.custom_field_option_ids.to_set

    errors.add(:distribution, 'options must be a subset of the options of the associated custom field.')
```

adessy replied:
Here, i'd rather keep it as it is. I think it makes the link between the error message and the condition more explicit.

koendehondt commented on back/engines/commercial/user_custom_fields/app/serializers/user_custom_fields/web_api/v1/ref_distribution_serializer.rb:
```diff
@@ -0,0 +1,20 @@
+
+        attribute :distribution, &:normalized_distribution
+
+        # TODO: Improve performance
```
Please resolve this to-do.

adessy replied:
Just removed the comment from now. I don't expect big performance issues here. I was a bit annoyed the serialization triggers a query to the DB which can be an issue when serializing large collection of objects. But it should not be the case here.

adessy replied:
In the end, I added `values` as an association and used `includes` to eager-load them. Not perfect, but still a bit better.

koendehondt commented on back/engines/commercial/user_custom_fields/config/routes.rb:
```diff
@@ -9,11 +9,13 @@
             patch 'reorder', on: :member
           end
+
+          resource :reference_distribution, controller: 'ref_distributions', only: %i[show create destroy]
```
Consider renaming `RefDistributionsController` to `ReferenceDistributionsController`. Then  the default Rails conventions are used and `controller: 'ref_distributions' is not required here.

adessy replied:
Yes, I pondered a bit about it. But then, it would also make sense to rename the model and the db tables, defeating the original goal of avoiding very lonnnng names (`UserCustomFields::Representativeness::ReferenceDistribution`). I ended up applying the following decision rule: anything that an implementation detail is written as `RefDistribution` and anything that's part of an interface with the outside world is written as `reference_distribution`. We could very much argue that the web/front-end API is external...

After having written this down, I would actually tend to rename the endpoint to ref_distribution instead of renaming the controller. (I can't help but note that as convenient as they are, rails conventions are actually encouraging us to expose implementation details in this case.)

koendehondt replied:
I do not know yet what the CL convention is, but why is it necessary to have long names? The name `user_custom_fields_representativeness_ref_distributions` is a very long table name and I wonder whether it is useful. Is it so long to express the relationship with user custom fields and the representativeness? Wouldn't `reference_distributions` be enough?

adessy replied:
Well, I don't think there is a guideline for that. Personally, I thought it was a reasonable thing to do to namespace the tables similarly to the active record models. I don't think it's particularly useful in this case, but I don't think it's hurting either, is it? Because in most cases, the table name is not really something you would write explicitly in the code (except for more complex queries). Is it more the case with Sequel?

But I'm not sure to understand the link between your original comment and the table name.

koendehondt replied:
> Is it more the case with Sequel?

No.

> But I'm not sure to understand the link between your original comment and the table name.

My original comment was about the controller name, not about the table name 😄 . Let's close this discussion.

koendehondt commented on back/engines/commercial/user_custom_fields/spec/acceptance/ref_distributions_spec.rb:
```diff
@@ -0,0 +1,125 @@
+      end
+    end
+
+    context 'when the custom field already have a distribution' do
```
```suggestion
    context 'when the custom field already has a distribution' do
```

koendehondt commented on back/engines/commercial/user_custom_fields/spec/acceptance/ref_distributions_spec.rb:
```diff
@@ -0,0 +1,125 @@
+require 'rails_helper'
+require 'rspec_api_documentation/dsl'
+
+resource 'Representativeness reference distributions' do
```
I know that it was not done in the past, but in `citizenlab/back/engines/commercial/content_builder/spec/acceptance/layouts_spec.rb` we have added examples to describe what happens when authorization fails. Without those examples, the API user does not know which http status will be returned. This is not a blocker for this PR. I just wanted to share it. Probably at some point we need a discussion about it at the CDM.

adessy replied:
Good catch! Actually, this is something I usually do, a bit differently though: `engines/commercial/insights/spec/acceptance/views_spec.rb`. I will add them.

koendehondt reviewed (changes requested):
Given that the ticket (CL-610) does not include any information on what had to be implemented unless what the title of the ticket says, it is hard for me to assess whether this PR implements what has been requested. In particular, I do not know what the controller should return.
There are some to-dos in the code of which I think they should be resolved before approving this PR. Otherwise they may stay to-dos forever. Apart from that, I have some suggestions to improve the code.

adessy commented:
Thanks for the swift review, @koendehondt .

> Given that the ticket ([CL-610](https://citizenlab.atlassian.net/browse/CL-610)) does not include any information on what had to be implemented unless what the title of the ticket says, it is hard for me to assess whether this PR implements what has been requested. In particular, I do not know what the controller should return.

Hence, the lengthy description in the PR itself. I always try to provide enough context to make the PR self-contained (either via the PR description, or the ticket). But when something is not clear enough, feel free to reach out to me before diving further in.

adessy committed:
Update back/engines/commercial/user_custom_fields/spec/acceptance/ref_distributions_spec.rb

Co-authored-by: Koen De Hondt <koen@dehondt.dev>

koendehondt commented:
> I always try to provide enough context to make the PR self-contained (either via the PR description, or the ticket)

👍 I do too 😃

adessy committed:
CL-610 Merge branch 'master' into CL-610/representativeness-reference-distribution

# Conflicts:
#	back/engines/commercial/user_custom_fields/config/routes.rb
#	back/engines/commercial/user_custom_fields/lib/user_custom_fields/engine.rb
#	back/engines/commercial/user_custom_fields/lib/user_custom_fields/feature_specification.rb
#	back/engines/commercial/user_custom_fields/spec/lib/settings_spec.rb

adessy committed:
CL-610 Merge remote-tracking branch 'origin/CL-610/representativeness-reference-distribution' into CL-610/representativeness-reference-distribution

adessy committed:
CL-610 Extract code into a method to remove the need for a comment

From a PR review suggestion: "If a comment is needed to explain a piece of code, abstraction is needed. Using expressive variable names and extracting code to new methods are the ways to achieve abstraction during day to day coding. By doing so, the code will be easier to read."

adessy committed:
CL-610 Remove TODO comment

I don't expect big performance issues here. I was a bit annoyed the serialization triggers a query to the DB which can be an issue when serializing large collection of objects. But it should not be the case here.

adessy committed:
CL-610 Implement ref. distribution values as an association

The values associated to a reference distribution can now be eager loaded using the association.

adessy committed:
CL-610 Add missing frozen-string-literal comment

adessy committed:
CL-610 Merge remote-tracking branch 'origin/CL-610/representativeness-reference-distribution' into CL-610/representativeness-reference-distribution

adessy committed:
CL-610 Add examples to test unauthorized requests

adessy committed:
CL-610 Test policy for reference distributions

adessy committed:
CL-610 Clarify TODO comment [skip ci]

adessy commented:
Hey @koendehondt, I tried to address your comments. Besides that, I just added some specs for `RefDistributionPolicy`.

adessy committed:
CL-610 Fix Rubocop offenses

adessy committed:
CL-610 Rename `normalized_distribution` method with a better name

adessy committed:
CL-610 Merge branch 'master' into CL-610/representativeness-reference-distribution

adessy commented:
Hi @koendehondt, just checking bf merging, are we good on this one?

adessy committed:
CL-610 Merge branch 'master' into CL-610/representativeness-reference-distribution

koendehondt commented on back/engines/commercial/user_custom_fields/app/controllers/user_custom_fields/web_api/v1/ref_distributions_controller.rb:
```diff
@@ -0,0 +1,59 @@
+          ref_distribution.save!
+        end
+
+        side_fx.after_create(ref_distribution, current_user)
```
`before_create` is missing from this method.

adessy replied:
I don't really see the issue here since there are no `before_create` effects in this resource. Of course, we might add some in the future, but I'd rather not anticipate and clutter the code with method calls that do nothing.

koendehondt replied:
I see that `Representativeness::SideFxRefDistributionService` inherits from `BaseSideFxService`, which does not implement `before_create`, so my comment was not valid. I expected `before_create` to be in the superclass, because it is present in similar classes, like `SideFxCommentService`. Apparently not all classes implement the side effects framework correctly.

So I will resolve this conversation.

koendehondt commented on back/engines/commercial/user_custom_fields/app/controllers/user_custom_fields/web_api/v1/ref_distributions_controller.rb:
```diff
@@ -0,0 +1,59 @@
+
+      def destroy
+        reference_distribution.destroy!
+        side_fx.after_destroy(reference_distribution, current_user)
```
`before_destroy` is missing from this method.

koendehondt replied:
This is similar to `before_create`. `before_destroy` does not exist on the side fx class.

koendehondt commented on back/engines/commercial/user_custom_fields/app/controllers/user_custom_fields/web_api/v1/ref_distributions_controller.rb:
```diff
@@ -0,0 +1,59 @@
+      end
+
+      def side_fx
+        Representativeness::SideFxRefDistributionService.new
```
Consider memoizing the result, because this method will be called twice from actions with `before_xxx` and `after_xxx`.

adessy replied:
Well, indeed if I add a call to `before_xxx`. In any case, I'll add it as it's low cost and makes the code more future-proof.

koendehondt commented on back/engines/commercial/user_custom_fields/app/models/user_custom_fields/patches/custom_field.rb:
```diff
@@ -0,0 +1,16 @@
+module UserCustomFields::Patches::CustomField
+  def self.included(base)
+    base.class_eval do
+      # TODO: Refactor CustomFields to use STI with two types: UserCustomField & FormCustomField
```
I would write "TODO tech debt:" at the beginning of this comment, because for the reader it is not clear whether it is a to-do for you, or a general to-do, i.e. tech debt.

adessy replied:
Changed it to `# TODO: (tech debt) ...` not to upset rubocop.

koendehondt commented on back/engines/commercial/user_custom_fields/app/models/user_custom_fields/representativeness/ref_distribution.rb:
```diff
@@ -0,0 +1,47 @@
+
+  def validate_distribution_options
+    return if custom_field.blank?
+    return if distribution.blank?
```
```suggestion
    return if custom_field.blank? || distribution.blank?
```

koendehondt commented on back/engines/commercial/user_custom_fields/spec/acceptance/ref_distributions_spec.rb:
```diff
@@ -0,0 +1,159 @@
+  end
+
+  post 'web_api/v1/users/custom_fields/:custom_field_id/reference_distribution' do
+    parameter :distribution, 'The distribution as (unnormalized) counts', required: true
```
```suggestion
    parameter :distribution, 'The distribution as (non normalized) counts', required: true
```

adessy replied:
`unnormalized` is the most common way of saying it in statistics, maybe bc "non normalized" looks a lot like "non normal" which is really different (but that's just a guess). I'll rework the description of the parameter though, as latest changes in the definition of the score (or lack of) add some more constraints to the counts.

koendehondt replied:
> is the most common way of saying it in statistics

I did not know. "unnormalized" does not seem to be an English word. That is why I made the suggestion.

koendehondt commented on back/engines/commercial/user_custom_fields/spec/acceptance/ref_distributions_spec.rb:
```diff
@@ -0,0 +1,159 @@
+
+    let(:custom_field_id) { custom_field.id }
+    let(:custom_field) do
+      create(:custom_field_select, resource_type: 'User').tap do |cf|
```
Please use at least 3 letters for a variable name. Soon 2 letters will not pass RuboCop. `custom_field` is even better, because the reader does not have to think about what the variable represents.

adessy replied:
sad 😢

koendehondt replied:
Why does it make you sad?

adessy replied:
Well, that's not something I would do on a regular basis but I think in some instances it's actually not hurting and even helps to make the code easier to read (e.g. collision with name in outer scope, longer lines -> multi-line block instead of single-line block).

koendehondt replied:
Code is more often read than written, so "even helps to make the code easier to read" is the main goal of using good names. The fact that we have longer lines is a side effect, but for the better.

adessy replied:
Allow me to disagree that all short names in all contexts are bad names :)

koendehondt commented on back/engines/commercial/user_custom_fields/spec/factories/user_custom_fields/representativeness/ref_distributions.rb:
```diff
@@ -0,0 +1,24 @@
+    custom_field do
+      # We use 'create' instead of 'association' because in any case we need the
+      # identifiers of custom field options to define the distribution.
+      create(:custom_field_select, resource_type: 'User').tap do |cf|
```
Same comment as before on `cf`.

koendehondt commented on back/engines/commercial/user_custom_fields/spec/models/user_custom_fields/representativeness/ref_distribution_spec.rb:
```diff
@@ -0,0 +1,33 @@
+describe UserCustomFields::Representativeness::RefDistribution do
+  subject(:ref_distribution) { build(:ref_distribution) }
+
+  describe 'factory' do
```
I have seen tests like this in other parts of the code base. I think it is strange. This test does not describe anything about the described class. It verifies that the factory produces a valid model, which should not be tested here imo.

adessy replied:
I understand your point, but consistency first I guess. That's something you might want to bring up during the CDM.

koendehondt replied:
At some point, I will :-)

koendehondt commented on back/engines/commercial/user_custom_fields/spec/models/user_custom_fields/representativeness/ref_distribution_spec.rb:
```diff
@@ -0,0 +1,33 @@
+  it { is_expected.to validate_uniqueness_of(:custom_field_id).case_insensitive }
+  it { is_expected.to validate_presence_of(:distribution) }
+
+  it 'validates that the distribution has at least 2 options', :aggregate_failures do
```
This test does not verify "at least 2 options". It verifies "exactly 2 options". So I think an extra expectation should be added for a distribution with 3 options.

adessy replied:
This spec is meant to be an (simplified) equivalent of something like:
```ruby
it { is_expected.to validate_length_of(:distribution).is_at_least(2).with_short_message('must have at least 2 options.') }
```
which doesn't work in this case because it tries with multiple test values that depends on the column type. In the case of a `jsonb`, those test values also include json arrays which are not valid/realistic distributions and it is freaking out other validation logic.

How would you go about that?

adessy replied:
Oops, sorry missed the part about:
> So I think an extra expectation should be added for a distribution with 3 options.

Actually, I chose to test the boundary here (2 -> OK, 1 -> KO). Why do you think adding an expectation with 3 options would help? 3 seems pretty arbitrary to me.

koendehondt replied:
Without the case for 3, I can change the validation to verify for exactly 2 options, like `validates :distribution, presence: true, length: { is: 2, message: 'must have exactly 2 options.' }`,  and the test would still pass.

adessy replied:
Devil's advocate: if I follow the same kind of reasoning, I can change the validation to verify for either 2 or 3 options: `validates :distribution, presence: true, length: {minimum: 2, maximum: 3}`. So, we would have to add an expectation to check that 4 options is allowed. Repeat the induction step, and now we need an expectation for all natural numbers >= 2.

koendehondt replied:
You are right. It never ends 😄 I will consider it personal preference and close the conversation.

koendehondt commented on back/engines/commercial/user_custom_fields/spec/policies/user_custom_fields/representativeness/ref_distribution_policy_spec.rb:
```diff
@@ -0,0 +1,33 @@
+
+require 'rails_helper'
+
+RSpec.describe UserCustomFields::Representativeness::RefDistributionPolicy, type: :policy do
```
Why is `type: :policy` required here?

adessy replied:
It is not 🙈

koendehondt commented on back/engines/commercial/user_custom_fields/spec/policies/user_custom_fields/representativeness/ref_distribution_policy_spec.rb:
```diff
@@ -0,0 +1,33 @@
+RSpec.describe UserCustomFields::Representativeness::RefDistributionPolicy, type: :policy do
+  subject(:policy) { described_class.new(user, ref_distribution) }
+
+  let_it_be(:ref_distribution) { create(:ref_distribution) }
```
Why don't you use `let` instead of `let_it_be" in this file?

adessy replied:
It makes the tests faster.

koendehondt replied:
How much faster? Why don't we use it in all specs if it makes the tests faster?

adessy replied:
In this case, I think the gain is around 20%. But it greatly depends on the test file. It's a pretty recent addition to the code base (around September 2021, I'd say) and I don't think everyone knows about it or feels comfortable enough using it (we cannot always use this trick).

koendehondt replied:
Seems to be a topic for the CDM :-)

adessy replied:
https://docs.google.com/document/d/1zcqyfubUoWikpzf1PzLcTEFI9MbAhZ9S5FbMqkkcQMc/edit?skip_itp2_check=true#heading=h.cc5vwaqbjg57 🙂

koendehondt reviewed (changes requested):
I have some extra review comments.

adessy committed:
Update back/engines/commercial/user_custom_fields/app/models/user_custom_fields/representativeness/ref_distribution.rb

Co-authored-by: Koen De Hondt <koen@dehondt.dev>

adessy committed:
CL-610 Remove useless rspec metadata

adessy committed:
CL-610 Merge remote-tracking branch 'origin/CL-610/representativeness-reference-distribution' into CL-610/representativeness-reference-distribution

adessy committed:
CL-610 Make parameter names longer

Soon 2 letters will not pass RuboCop.

adessy committed:
CL-610 Change TODO as suggested in review

> I would write "TODO tech debt:" at the beginning of this comment, because for the reader it is not clear whether it is a to-do for you, or a general to-do, i.e. tech debt.

adessy committed:
CL-610 Memoize side-effect service

adessy committed:
CL-610 Improve documentation for POST .../ref_distribution

adessy commented:
@koendehondt, are we good now? 😇

koendehondt reviewed (approved):
It involved some conversations, but I think it is ready now.

adessy committed:
CL-610 Merge branch 'master' into CL-610/representativeness-reference-distribution

adessy committed:
CL-610 Fix rubocop offense
